tests/view/treewalker.js

```diff
@@ -0,0 +1,628 @@
+				expectValue( value, expected[ i ] );
+				i++;
+			}
+
+			expect( i ).to.equal( expected.length );
+		} );
+
+		it( 'should provide iterator interface which BACKWARD direction', () => {
+			let iterator = new TreeWalker( { startPosition: rootEnding, direction: 'BACKWARD' } );
+			let i = expected.length;
+
+			for ( let value of iterator ) {
+				expectValue( value, expected[ --i ], { direction: 'BACKWARD' } );
+			}
+
+			expect( i ).to.equal( 0 );
+		} );
+
+		it( 'should start iterating at the startPosition witch is not a root bound', () => {
+			let iterator = new TreeWalker( { startPosition: new Position( root, 1 ) } );
+			let i = 2;
+
+			for ( let value of iterator ) {
+				expectValue( value, expected[ i ] );
+				i++;
+			}
+
+			expect( i ).to.equal( expected.length );
+		} );
+
+		it( 'should start iterating at the startPosition witch is not a root bound, going backward', () => {
+			let expected = [
+				{ type: 'ELEMENT_START', item: img1 },
+				{ type: 'ELEMENT_END', item: img1 }
+			];
+
+			let iterator = new TreeWalker( { startPosition: new Position( root, 1 ), direction: 'BACKWARD' } );
+			let i = expected.length;
+
+			for ( let value of iterator ) {
+				expectValue( value, expected[ --i ], { direction: 'BACKWARD' } );
+			}
+
+			expect( i ).to.equal( 0 );
+		} );
+	} );
+
+	describe( 'iterate trough the range `boundary`', () => {
+		describe( 'range starts between elements', () => {
+			let expected, range;
+
+			before( () => {
+				expected = [
+					{ type: 'ELEMENT_START', item: paragraph },
+					{ type: 'ELEMENT_START', item: bold },
+					{ type: 'TEXT', text: 'ba' },
+					{ type: 'ELEMENT_END', item: bold },
+					{ type: 'TEXT', text: 'r' },
+					{ type: 'ELEMENT_START', item: img2 },
+					{ type: 'ELEMENT_END', item: img2 }
+				];
+
+				range = Range.createFromParentsAndOffsets( root, 1, paragraph, 3 );
+			} );
+
+			it( 'should iterating over the range', () => {
+				let iterator = new TreeWalker( { boundaries: range } );
+				let i = 0;
+
+				for ( let value of iterator ) {
+					expectValue( value, expected[ i ] );
+					i++;
+				}
+
+				expect( i ).to.equal( expected.length );
+			} );
+
+			it( 'should iterating over the range going backward', () => {
+				let iterator = new TreeWalker( { boundaries: range, direction: 'BACKWARD' } );
+				let i = expected.length;
+
+				for ( let value of iterator ) {
+					expectValue( value, expected[ --i ], { direction: 'BACKWARD' } );
+				}
+
+				expect( i ).to.equal( 0 );
+			} );
+		} );
+
+		describe( 'range starts inside the text', () => {
+			let expected, range;
+
+			before( () => {
+				expected = [
+					{ type: 'TEXT', text: 'a' },
+					{ type: 'ELEMENT_END', item: bold },
+					{ type: 'TEXT', text: 'r' },
+					{ type: 'ELEMENT_START', item: img2 },
+					{ type: 'ELEMENT_END', item: img2 }
+				];
+
+				range = Range.createFromParentsAndOffsets( textBa, 1, paragraph, 3 );
+			} );
+
+			it( 'should return part of the text', () => {
```

Comment:
This tests does not catch the bug that single letters are returned when you start in the middle of the text node.